executor/index_lookup_hash_join.go

```diff
@@ -0,0 +1,449 @@
+		copiedRanges = append(copiedRanges, ran.Clone())
+	}
+	iw := &indexHashJoinInnerWorker{
+		innerWorker: innerWorker{
+			innerCtx:      e.innerCtx,
+			outerCtx:      e.outerCtx,
+			ctx:           e.ctx,
+			executorChk:   chunk.NewChunkWithCapacity(e.innerCtx.rowTypes, e.maxChunkSize),
+			indexRanges:   copiedRanges,
+			keyOff2IdxOff: e.keyOff2IdxOff,
+		},
+		taskCh:            taskCh,
+		joiner:            e.joiner,
+		joinChkResourceCh: e.joinChkResourceCh[workerID],
+		resultCh:          e.resultCh,
+		matchedOuterPtrs:  make([][]byte, 0, 8),
+	}
+	iw.finished.Store(false)
+	return iw
+}
+
+func (iw *indexHashJoinInnerWorker) run(ctx context.Context) {
+	var task *indexHashJoinTask
+	joinResult, ok := iw.getNewJoinResult(ctx)
+	if !ok {
+		return
+	}
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case task, ok = <-iw.taskCh:
+		}
+		if !ok {
+			break
+		}
+		if task.err != nil {
+			joinResult.err = task.err
+			break
+		}
+		err := iw.handleTask(ctx, task, joinResult)
+		if err != nil {
+			joinResult.err = err
+			break
+		}
+	}
+	if joinResult.err != nil || (joinResult.chk != nil && joinResult.chk.NumRows() > 0) {
+		select {
+		case iw.resultCh <- joinResult:
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func (iw *indexHashJoinInnerWorker) getNewJoinResult(ctx context.Context) (*indexHashJoinResult, bool) {
+	joinResult := &indexHashJoinResult{
+		src: iw.joinChkResourceCh,
+	}
+	ok := true
+	select {
+	case joinResult.chk, ok = <-iw.joinChkResourceCh:
+	case <-ctx.Done():
+		return nil, false
+	}
+	return joinResult, ok
+}
+
+func (iw *indexHashJoinInnerWorker) buildHashTableForOuterResult(ctx context.Context, task *indexHashJoinTask) {
+	var (
+		keyBuf = make([]byte, 0, 64)
+		valBuf = make([]byte, 8)
+		err    error
+	)
+	for rowIdx, numRows := 0, task.outerResult.NumRows(); rowIdx < numRows; rowIdx++ {
+		var hasNull bool
+		if task.outerMatch != nil && !task.outerMatch[rowIdx] {
+			continue
+		}
+		row := task.outerResult.GetRow(rowIdx)
+		keyColIdx := iw.outerCtx.keyCols
+		for _, i := range keyColIdx {
+			if row.IsNull(i) {
+				hasNull = true
+				break
+			}
+		}
+		if hasNull {
+			continue
+		}
+		keyBuf, err = codec.HashChunkRow(iw.ctx.GetSessionVars().StmtCtx, keyBuf[:0], row, iw.outerCtx.rowTypes, keyColIdx)
+		if err != nil {
+			iw.finished.Store(true)
+			result, ok := iw.getNewJoinResult(ctx)
+			if !ok {
+				return
+			}
+			result.err = err
+			select {
+			case iw.resultCh <- result:
+			case <-ctx.Done():
+			}
+		}
+		*(*int)(unsafe.Pointer(&valBuf[0])) = rowIdx
+		task.lookupMap.Put(keyBuf, valBuf)
+	}
+	return
+}
+
+func (iw *indexHashJoinInnerWorker) fetchInnerResults(ctx context.Context, task *lookUpJoinTask) error {
+	lookUpContents, err := iw.constructLookupContent(task)
+	if err != nil {
+		return err
+	}
+	lookUpContents = iw.sortAndDedupLookUpContents(lookUpContents)
+	if iw.finished.Load().(bool) {
+		return nil
+	}
+	return iw.innerWorker.fetchInnerResults(ctx, task, lookUpContents)
```

Comment:
If the task would have a large set of results, may we do canceling inside `iw.innerWorker.fetchInnerResults` for releasing resources earlier?

Reply:
Maybe we can set the `sessVars.Killed` instead of using iw.finished ?

Reply:
`sessVars.Killed` is anthor signal, we may consider it later.

The only usage of `iw.finished` is here,  https://github.com/pingcap/tidb/pull/8661/files#diff-26b1f580c7020b9151e1fb5f87d5a44dR334
Perharps we could simply call `cancelFunc` here for exiting,  and handle `ctx.Done` in `iw.innerWorker.fetchInnerResults`.

Reply:
I tried, but do not find a place to check the `ctx.Done`.

Reply:
How about checking `ctx.Done` before every `Next` https://github.com/pingcap/tidb/blob/master/executor/index_lookup_join.go#L593